### PR TITLE
Unify multiple Gemfile and Gemfile.lock

### PR DESCRIPTION
**Pull request by kenhys**
Since bundler 2.2.0, automultiplatform feature is enabled again.

See https://github.com/rubygems/bundler/pull/7580 in details.

### CODE REVIEW

**Comment by kenhys** [comment]
```
PS C:\opt\td-agent\share> ..\bin\td-agent-gem.bat list | Select-String  bundler

bundler (2.2.0.rc.1, default: 2.1.4)
```

```
PS C:\opt\td-agent\share> ..\bin\fluentd --gemfile .\Gemfile
Warning: the running version of Bundler (2.1.4) is older than the version that created the lockfile (2.2.0.rc.1). We suggest you to upgrade to the version that created the lockfile by running `gem install bundler:2.2.0.rc.1 --pre`.
Fetching gem metadata from https://rubygems.org/............
Fetching gem metadata from https://rubygems.org/.
Could not find gem 'win32-api (= 1.9.1)' in any of the gem sources listed in
your Gemfile.
```

**Comment by kenhys** [comment]
bundler 2.2.0.rc1

* bundle lock --add-platform doesn't work `if RUBY_PLATFORM` clause exists.
* Even though comment out `if` clause from Gemfile, `win22-api = (1.9.1) is not resolved.

**Comment by kenhys** [comment]
On the other hand,  on windows:

```
PS C:\work\td-agent-builder.work\td-agent> bundle -v
Bundler version 2.2.0.rc.1
PS C:\work\td-agent-builder.work\td-agent> bundle update --all
Fetching https://github.com/fluent/fluentd
Fetching gem metadata from https://rubygems.org/............
Fetching gem metadata from https://rubygems.org/.
Could not find gem 'win32-api (= 1.9.1)' in any of the gem sources listed in
your Gemfile or in gems cached in
C:/work/td-agent-builder.work/td-agent/downloads.
```

Hmm, bundler 2.2.0.rc1 is not support universal gem yet?

**Comment by cosmo0920** [comment]
I'd published win32-api non-universal gem. Could you try Gemfile unification with `platforms`?

**Comment by kenhys** [comment]
> I'd published win32-api non-universal gem. Could you try Gemfile unification with `platforms`?

Thanks! I'll try it.

File: td-agent/Gemfile.lock [hunk 1]
```diff
@@ -4,7 +4,6 @@ GIT
   ref: 2f80f6cf0a4aa44b9795cd6a66894a6b8fa70d06
   specs:
     fluentd (1.11.2)
-      certstore_c (~> 0.1.7)
```
Comment (cosmo0920) [inline td-agent/Gemfile.lock:7]: Can we include windows specific Fluentd dependencies in Gemfile.lock?

File: td-agent/Gemfile.lock [hunk 2]
```diff
@@ -13,10 +12,6 @@ GIT
       strptime (>= 0.2.2, < 1.0.0)
       tzinfo (>= 1.0, < 3.0)
       tzinfo-data (~> 1.0)
-      win32-event (~> 0.6.3)
-      win32-ipc (~> 0.7.0)
-      win32-service (~> 2.1.5)
-      windows-pr (~> 1.2.6)
```
Comment (cosmo0920) [inline td-agent/Gemfile.lock:19]: ditto.

File: td-agent/Gemfile.lock (hunk 1 above)
Comment (cosmo0920) [reply td-agent/Gemfile.lock:7]: Ah! They are already included.

**Comment by cosmo0920** [comment]
Umm..., serverspec complains failure to gem installation testcases:
https://github.com/fluent-plugins-nursery/td-agent-builder/pull/190/checks?check_run_id=1064622440#step:6:5840
Should we check gem platform and reject gem installation tests properly?

**Comment by kenhys** [comment]
On windows, we need to bundle update for windows specific gem in Gemfile.lock

**Comment by kenhys** [comment]
centos6: `console-1.9.0 requires ruby version >= 2.5, which is incompatible with the current version, ruby 2.4.6p354`

**Comment by kenhys** [comment]
centos7: 

```
opt/td-agent/bin/gem install --no-document --bindir /root/rpmbuild/BUILDROOT/td-agent-4.0.1-1.el7.x86_64/opt/td-agent/bin /root/rpmbuild/BUILD/td-agent-4.0.1/td-agent/downloads/bundler-2.2.0.rc.1.gem
rake aborted!
TypeError: no implicit conversion of Array into String
/root/rpmbuild/BUILD/td-agent-4.0.1/td-agent/Rakefile:702:in `gem_install'
/root/rpmbuild/BUILD/td-agent-4.0.1/td-agent/Rakefile:732:in `install_gems'
/root/rpmbuild/BUILD/td-agent-4.0.1/td-agent/Rakefile:352:in `block (2 levels) in define'
/opt/rh/rh-ruby26/root/usr/share/gems/gems/rake-12.3.2/exe/rake:27:in `<top (required)>'
Tasks: TOP => build:all => build:licenses => build:gems => build:ruby_gems
```

**Comment by kenhys** [comment]
centos8: rpmlint has failed.

```
0 packages and 0 specfiles checked; 0 errors, 15 warnings.                                                                                                   
Traceback (most recent call last):                                                                                                                           
  File "/usr/bin/rpmlint", line 378, in <module>                                                                                                             
    main()                                                                                                                                                   
  File "/usr/bin/rpmlint", line 166, in main                                                                                                                 
    runChecks(pkg)                                                                                                                                           
  File "/usr/bin/rpmlint", line 223, in runChecks                                                                                                            
    check.check(pkg)                                                                                                                                         
  File "/usr/share/rpmlint/ZipCheck.py", line 43, in check                                                                                                   
    badcrc = z.testzip()                                                                                                                                     
  File "/usr/lib64/python3.6/zipfile.py", line 1293, in testzip                                                                                              
    with self.open(zinfo.filename, "r") as f:                                                                                                                
  File "/usr/lib64/python3.6/zipfile.py", line 1429, in open                                                                                                 
    "required for extraction" % name)                                                                                                                        
RuntimeError: File 'file1.txt' is encrypted, password required for extraction                                                                                
cd -                                                                                                                                                         
cd -                                      
```

**Comment by kenhys** [comment]
focal: stall with lintian (encyrypted zip)

```
dpkg-buildpackage: info: full upload (original source is included)
Now running lintian --profile td-agent/ubuntu td-agent_4.0.1-1_amd64.changes ...
warning: the authors of lintian do not recommend running it with root privileges!

```

**Comment by kenhys** [comment]
windows: bundler 2.2.0.rc.1 is installed, but not default.

**Comment by kenhys** [comment]
> centos8: rpmlint has failed.

it should be installed under opt/td-agent/lib/ruby/2.7.0.

**Comment by kenhys** [comment]
rebased with the recent master branch. (WIP)

**Comment by kenhys** [comment]
we need to install bundler before executing `bundle install --gemfile`.

**Comment by kenhys** [comment]
> Unsupposed gemdir: /build/td-agent-4.0.1/debian/tmp/opt/td-agent/lib/ruby/gems/2.7.0 (expected: /opt/td-agent/lib/ruby/gems/2.7.0)

It will be fixed.

**Comment by kenhys** [comment]
```
  dh_strip_nondeterminism
strip-nondeterminism: debian/td-agent/opt/td-agent/lib/ruby/2.7.0/gems/rubyzip-1.3.0/test/data/zipWithEncryption.zip: ignoring encrypted zip file
        Normalized debian/td-agent/opt/td-agent/lib/ruby/2.7.0/gems/rubyzip-1.3.0/test/data/zipWithEncryption.zip
   dh_compress
   dh_fixperms
   dh_missing
   dh_strip
```

ruby-zip causes stall on focal, it is already known issue, but path seems a bit strange.

**Comment by kenhys** [comment]
> ```
>   dh_strip_nondeterminism
> strip-nondeterminism: debian/td-agent/opt/td-agent/lib/ruby/2.7.0/gems/rubyzip-1.3.0/test/data/zipWithEncryption.zip: ignoring encrypted zip file
>         Normalized debian/td-agent/opt/td-agent/lib/ruby/2.7.0/gems/rubyzip-1.3.0/test/data/zipWithEncryption.zip
>    dh_compress
>    dh_fixperms
>    dh_missing
>    dh_strip
> ```
> 
> ruby-zip causes stall on focal, it is already known issue, but path seems a bit strange.
> ruby-zip is pulled td => zip-zip => rubyzip dependency.

rubyzip => /lib/ruby/2.7.0/gems/rubyzip-1.3.0
other gems => /lib/ruby/gems/ruby/gems/foobar-xxx

**Comment by cosmo0920** [comment]
Hmm..., macOS build is failing certificates issue...:
https://github.com/fluent-plugins-nursery/td-agent-builder/runs/1079354029?check_suite_focus=true#step:3:35963

It's strange building failure.

**Comment by cosmo0920** [comment]
I'd fixed rdkafka gem building failure on https://github.com/fluent-plugins-nursery/td-agent-builder/pull/190/commits/a887fe225be28c82ed10801d3c1fca77e1b28ca3.

**Comment by kenhys** [comment]
deb and rpm will be built.

**Comment by kenhys** [comment]
> deb and rpm will be built.

Yay, bulid process has passed, then need to fix test cases

**Comment by kenhys** [comment]
But, some installed gem is not recognized by td-agent-gem.

**Comment by kenhys** [comment]
```
lib/ruby/2.7.0/gems/ not recognized gem
lib/ruby/gems/2.7.0/gems/*  recognized gem
```

**Comment by kenhys** [comment]
Fixed windows test case about path to Gemfile.lock

**Comment by kenhys** [comment]
`bundle install` installs gem under `lib/ruby/2.7.0/gems/`

**Comment by kenhys** [comment]
For example,  `bundle install --gemfile --path /tmp/opt` then gem file are installed under `/tmp/opt/ruby/2.7.0/gems`. it expects `/tmp/opt/ruby/gems/2.7/0/gems` but it is not.

**Comment by kenhys** [comment]
Just an idea: check Gemfile dependency which gem should be installed explicitly.

```
Bundler::Definition.build("td-agent/Gemfile", "td-agent/Gemfile.lock", false).dependencies.each do |spec|
  p "#{spec.name} => #{spec.platforms}"
end
```

```
"nokogiri => []"
"ffi => [:mingw, :x64_mingw]"
"ffi-win32-extensions => [:mingw, :x64_mingw]"
"win32-api => [:mingw, :x64_mingw]"
"win32-event => [:mingw, :x64_mingw]"
"win32-ipc => [:mingw, :x64_mingw]"
"win32-service => [:mingw, :x64_mingw]"
"windows-api => [:mingw, :x64_mingw]"
"windows-pr => [:mingw, :x64_mingw]"
"winevt_c => [:mingw, :x64_mingw]"
"win32-eventlog => [:mingw, :x64_mingw]"
"fluent-plugin-parser-winevt_xml => [:mingw, :x64_mingw]"
"fluent-plugin-windows-eventlog => [:mingw, :x64_mingw]"
"rdkafka => [:ruby]"
"systemd-journal => [:ruby]"
"fluent-plugin-systemd => [:ruby]"
```

**Comment by kenhys** [comment]
Found that 2.2.0 is not installed as expected, so I'm fixing...

**Comment by kenhys** [comment]
TODO:

windows: fix protocol-http2 0.14.1 error see https://github.com/socketry/protocol-http2/issues/6
xenial: fix zookeeper server error.

File: serverspec/windows/td-agent.rb [hunk 3]
```diff
@@ -26,11 +26,14 @@ def check_is_installed_by_gem(name, version=nil, gem_binary="gem")
...
   lock_path = File.join(File.dirname(File.dirname(File.dirname(__FILE__))),
-                        "gemfiles/windows/Gemfile.lock")
-  parser = Bundler::LockfileParser.new(Bundler.read_file(lock_path))
-  parser.specs.each do |spec|
-    describe package("#{spec.name}") do
-      it { should be_installed.by("gem").with_version(spec.version) }
+                        "td-agent/Gemfile.lock")
+  gem_path = File.join(File.dirname(lock_path),
+                       File.basename(lock_path, ".lock"))
+  Bundler::Definition.build(gem_path, lock_path, false).dependencies.each do |spec|
```
Comment (cosmo0920) [inline serverspec/windows/td-agent.rb:32]: Bundler::Definition should not handle `#version`:
https://docs.ruby-lang.org/en/2.6.0/Bundler/Definition.html

Comment (kenhys) [reply serverspec/windows/td-agent.rb:33]: Thanks!

**Comment by kenhys** [comment]
All Green!

File: td-agent/Gemfile [hunk 4]
```diff
@@ -50,24 +53,24 @@ gem "fluent-plugin-record-modifier", "2.1.0"
...
+gem "windows-pr", "1.2.6", platforms: windows_platforms
+gem "winevt_c", "0.8.1", platforms: windows_platforms
+gem "win32-eventlog", "0.6.7", platforms: windows_platforms
+gem "fluent-plugin-parser-winevt_xml", "0.2.2", platforms: windows_platforms
+gem "fluent-plugin-windows-eventlog", "0.7.0", platforms: windows_platforms
+
+not_windows_platforms = [:ruby]
+gem "rdkafka", "0.8.0", platforms: not_windows_platforms
+gem "systemd-journal", "1.3.3", platforms: not_windows_platforms
+gem "fluent-plugin-systemd", "1.0.2", platforms: not_windows_platforms
```
Comment (cosmo0920) [inline td-agent/Gemfile:73]: nitpick: Systemd is Linux only. But macOS users seems to be a few.
It is not issued in the near future.

**Review by cosmo0920 — APPROVED** [review]
LGTM.

Workaround code such as bundler 2.2.0.rc and protocol-http2 will be removed for the future version?

**Comment by kenhys** [comment]
> Workaround code such as bundler 2.2.0.rc and protocol-http2 will be removed for the future version?

Yes, It should be removed in the future.

**Pull request merged**